src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/FindAllReferences/FindAllReferencesEndpoint.cs

```diff
@@ -0,0 +1,157 @@
+
+        protected override Task<IDelegatedParams?> CreateDelegatedParamsAsync(ReferenceParamsBridge request, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var documentContext = requestContext.GetRequiredDocumentContext();
+            return Task.FromResult<IDelegatedParams?>(new DelegatedPositionParams(
+                    documentContext.Identifier,
+                    projection.Position,
+                    projection.LanguageKind));
+        }
+
+        protected override async Task<VSInternalReferenceItem[]> HandleDelegatedResponseAsync(VSInternalReferenceItem[] delegatedResponse, ReferenceParamsBridge originalRequest, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var remappedLocations = new List<VSInternalReferenceItem>();
+
+            foreach (var referenceItem in delegatedResponse)
+            {
+                if (referenceItem?.Location is null || referenceItem.Text is null)
+                {
+                    continue;
+                }
+
+                // Temporary fix for codebehind leaking through
+                // Revert when https://github.com/dotnet/aspnetcore/issues/22512 is resolved
+                referenceItem.DefinitionText = FilterReferenceDisplayText(referenceItem.DefinitionText);
+                referenceItem.Text = FilterReferenceDisplayText(referenceItem.Text);
+
+                // Indicates the reference item is directly available in the code
+                referenceItem.Origin = VSInternalItemOrigin.Exact;
+
+                if (!_featureOptions.IsVirtualCSharpFile(referenceItem.Location.Uri) &&
+                    !_featureOptions.IsVirtualHtmlFile(referenceItem.Location.Uri))
+                {
+                    // This location doesn't point to a virtual file. No need to remap.
+                    remappedLocations.Add(referenceItem);
+                    continue;
+                }
+
+                var documentContext = requestContext.GetRequiredDocumentContext();
+                var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+                if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, referenceItem.Location.Range, out var documentRange))
+                {
+                    continue;
+                }
+
+                var razorDocumentUri = _featureOptions.GetRazorDocumentUri(referenceItem.Location.Uri);
+                referenceItem.Location.Uri = razorDocumentUri;
+                referenceItem.DisplayPath = razorDocumentUri.AbsolutePath;
+                referenceItem.Location.Range = documentRange;
+
+                remappedLocations.Add(referenceItem);
+            }
+
+            return remappedLocations.ToArray();
+        }
+
+        private static object? FilterReferenceDisplayText(object? referenceText)
```

Comment:
I think these are deep enough wizardry that they warrant some sort of a top-level explanation of what's being accomplished and how.